### ISSUE TITLE
BatchSpanProcessor silently drops spans when the queue is full but does not warn about it in the logs

### ISSUE THREAD

**Issue by tim-mwangi** [open-telemetry/opentelemetry-go#3264]
### Problem Statement

The BatchSpanProcessor - https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/batch_span_processor.go -has a `MaxQueueSize` defaulted to 2048 and when this is full, it silently drops the spans sent to it - https://github.com/open-telemetry/opentelemetry-go/blob/697d245b9ccf375c1d49e7492d36bceae80acfe4/sdk/trace/batch_span_processor.go#L416. However, it does not log a warning or notify the user in any way. The only way to know this happened is to turn on debug logging - https://github.com/open-telemetry/opentelemetry-go/blob/697d245b9ccf375c1d49e7492d36bceae80acfe4/sdk/trace/batch_span_processor.go#L268. Debug logging could be too noisy.

### Proposed Solution

It would be nice that when the `bsp.dropped > 0` a warning is logged to notify the user that they need to look into it and increase the `MaxQueueSize`.  

#### Alternatives

Debug logging which is noisy since it happens at the export period which is too short. So the logs could easily be flooded.

**Comment by fabianaferreira** [comment]
Hello! I'd like to work on that, if possible. It would be my first contribution to the project!

From my understanding, it would be enough to use warning logs when processing the queue. Does it make sense? Thanks!

**Comment by tim-mwangi** [comment]
Hi @fabianaferreira , yes. So at line https://github.com/open-telemetry/opentelemetry-go/blob/697d245b9ccf375c1d49e7492d36bceae80acfe4/sdk/trace/batch_span_processor.go#L268 split up the logging such that we still debug log the length of the batch queue and if `bsp.dropped > 0` we warn log it.

**Comment by Aneurysm9** [comment]
I'm not sure that logging at the warn level on every dropped span is appropriate.  This is a situation that is likely to occur under load and adding a substantial amount of logging may exacerbate that load.  The number of dropped spans is expected to be able to be reported as a metric.  While we can't utilize the metric API in trace components at this time, that should be the way forward here IMO.

**Comment by ryanericson** [comment]
@Aneurysm9 That logging is not going to log at every single dropped span. It's at the time of export so we're looking at every batch, in case you really mean log message occuring for every single dropped span.

This is also not different than what we do at https://github.com/open-telemetry/opentelemetry-go/blob/31661dd6bed7ca1b87954d980bd0f6adc6a5d4c2/exporters/otlp/internal/retry/retry.go#L107
In the OTLP exporter if max retry has occured or if error isn't retryable every failed export attempt is going to show up as warning too, and those are situations under load as well for sure. To me this ticket isn't asking for anything, it's just making batch processing error logging consistent with exporter.

But the point to be able to tell when a drop is happening (which is a big issue in many use cases). If we don't have metrics  or warn logs how would we know that drops are happening at that point? Can we have the warn logs until we have metrics ready?

**Comment by Aneurysm9** [comment]
The OTLP exporter is not logging there, it just returns an error.

**Comment by ryanericson** [comment]
> The OTLP exporter is not logging there, it just returns an error.

Right, but in the end that error goes to `otel.Handle` which by default logs to stderr. 
https://github.com/open-telemetry/opentelemetry-go/blob/1f5b159161e0ff54df4677284fa7d6718dc9186d/sdk/trace/batch_span_processor.go#L298

**Comment by prodion23** [comment]
I attempted this as @tim-mwangi / @ryanericson suggested, a little unsure about what log level we should use since `.Warn` just seems to overlap with error in this case & modifying info logging integer level could have wider implications? 

https://github.com/open-telemetry/opentelemetry-go/pull/3289

**Comment by dmathieu** [comment]
Yes, metrics are stable now. So emitting one would be the way to go.
Related (actually, both issues are duplicates): #5557.
